rfcs/require-by-string.md

```diff
@@ -0,0 +1,360 @@
+
+To require a Luau module under the current implementation, we must require it either by relative or absolute path:
+
+```lua
+-- Requiring Roact by absolute path
+local Roact = require("C:/LuauModules/Roact-v1.4.2")
+```
+
+If we wanted to change the location of this package on disk, we would have to modify all files that require the old path above, which essentially boils down to manual package management. This approach is tedious and prone to mistakes.
+
+To solve this, we introduce aliases in this RFC, which would allow for future package management software to maintain package paths and versions by directly modifying alias maps.
+
+This would also create simple and readable require statements for developers.
+
+```lua
+-- Suppose "@src" is an alias for the same directory as "../../../../../"
+
+-- Instead of this:
+local result = require("../../../../../HelperModules/graphing")
+
+-- We could have this:
+local result = require("@src/HelperModules/graphing")
+```
+
+## Design
+
+### String syntax
+
+Our require syntax should:
+
+- Allow a library written in Luau to be imported into the Roblox engine and "just work".
+- Support compile-time file resolution _where possible_ for type checking.
+- Be consistent across platforms and both inside and outside of Roblox.
+- Provide support for binding unambiguous aliases to relative and absolute paths.
+- Attempt to remain consistent with the existing Luau path syntax.
+
+For compatibility across platforms, we will automatically map `/` onto `\`.
+
+#### Relative paths
+
+Modules can be required relative to the requiring file's location in the filesystem (note, this is different from the current implementation, which evaluates all relative paths in relation to the current working directory).
+
+If we are trying to require a module called `MyModule.luau` in `C:/MyLibrary`:
+```lua
+-- From C:/MyLibrary/SomeModule.luau
+local MyModule = require("MyModule")
+
+-- Same as above, identical behavior
+local MyModule = require("./MyModule")
+
+-- From C:/MyLibrary/SubDirectory/SubModule.luau
+local MyModule = require("../MyModule")
+ 
+-- From C:/MyOtherLibrary/MainModule.luau
+local MyModule = require("../MyLibrary/MyModule")
+```
+
+Relative paths are chosen as the default path type, meaning that if a given path does not begin with a reserved prefix such as `/` or `@`, then it is considered relative to the requiring file's location. This is chosen for convenience, as libraries may include various internal dependencies that are easier to reference using relative paths.
+
+Relative paths can also begin with `./` or `../`, which denote the directory of the requiring file and its parent directory, respectively. When executing Luau code directly from a CLI (e.g. in the Luau CLI), this is relative to the current working directory.
```

Comment:
I think this is a little misleading, maybe this is a vestigial part of the notion of tying require paths to cwd - I think this proposal should not rely on CWD. My understanding is:
- require(path) always evaluates path relative to the path of the module that is calling require
- when the code that is calling require doesn't have an associated path (eg it was compiled via loadstring), relative require paths result in an error
- Luau CLI uses CWD to translate command line arguments that refer to file paths to full file paths; this is the only connection to CWD in the entire proposal, and it's specific to the CLI.

Specifically, `require("../foo")` evaluates `..` relative to the path of the outer module - not cwd - even in CLI.

You can (and should!) use module source (lua_getinfo -> lua_Debug::source) to evaluate the module path. When it begins with =/@ it's a path which can be used per logic above; when it's something else, it's loadstring'ed or otherwise not tied to paths.

One weird caveat is REPL should probably support require based on CWD... but it more or less works as REPL executed code currently gets a fake module name "=stdin" if I recall, so it runs as if code was written in a file called stdin...

Reply:
Right, I think this might be me not writing clearly (will update the RFC). When I wrote "executing Luau code directly from a CLI" I meant only the case when we call `require` directly in the REPL. In my local implementation that I'm working on, I have it as you've described: the "=stdin" special case is the only time that a relative path is evaluated relative to the CWD—otherwise, it's always evaluated relative to the requiring file. I will also add a note about code that is compiled via `loadstring`.